doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,562 @@
+     assume one would like to copy/move a shared memory located vector into a local
+     cache. If the allocator would change the container has to allocate another
+     chunk inside the shared memory for the data copy. This would use up a lot
+     of shared memory for local copies but the purpose of the shared memory is
+     to store data for zero copy communication.
+   * It would make it impossible for users to define local allocators in members
+     and use the compiler generated copy/move assignment operator.
+     ```cpp
+     struct MyData {
+       cxx::vector<int> value;
+     };
+
+     dataOnSharedMemory = subscriber->take();
+     MyData cache{.value = cxx:vector<int>(HeapAllocator)};
+     cache = dataOnSharedMemory; // would suddenly allocate shared memory
+     ```
+
+The draft we provide here is using the `cxx::vector` but the techniques
+described can be easily applied to the `cxx::string` or other cxx containers.
+
+```
++------------------------------------------------------+
+|template<T>                                           |
+|vector                                                |
+|                                                      |
+|  - vector() = default                                |
+|  - template<Allocator>                               |
+|    vector(Allocator & allocator);                    |
+|                                                      |
+|  - expected<AllocatorError> reserve(const uint64_t); |
+|  - void release();                                   |
+|                                                      |
+|  // remaining API is unchanged                       |
++------------------------------------------------------+
+                          ^
+                         /|\
+                          |
+       +------------------+--------------------------+
+       |template<T, Capacity>                        |
+       |vector : public vector<Type>                 |
+       |                                             |
+       |  (#1)                                       |
+       |  - vector(const vector<T> &)                |
+       |  - vector(vector<T> &&)                     |
+       |  - vector& operator=(const vector<T> &)     |
+       |  - vector& operator=(vector<T> &&)          |
+       |                                             |
+       |                                             |
+       |  // same API as current cxx::vector         |
+       |                                             |
+       |  # uint8_t m_data[Capacity]                 |
+       |  # RangeAllocator m_allocator               |
+       +---------------------------------------------+
+```
+All non stack based version of cxx containers will not have constructors similar
+to the STL. The only provided constructor has one argument which specifies the
+type of allocator one would like to use for that container. The default constructor
+shall construct a vector which has a capacity of 0.
```

Comment:
Can you please elaborate? Why can't we have typical c'tors with the allocator as last parameter with a default value, e.g. the heap allocator? This would be more user friendly.